### PR TITLE
Author and voc

### PR DESCRIPTION
**Pull request by wowtor**

### CODE REVIEW

File: authorship.yaml [hunk 1]
```diff
@@ -1,2 +1,3 @@
 data: frida/frida.sha256
+vocalise_data: frida/vocalize_results/Output_6-28-2021 2-19-04 PM_2019A-XVector-Adaptable-20F-512-OCMS-29May2019_XVector - PLDA_all_edits.csv
```
Comment (wowtor) [inline authorship.yaml]: Reference to a file which is not included. Could it be checked in or is it too big? (Simpler filename would be preferred.)

File: authorship/data.py [hunk 2]
```diff
@@ -57,13 +55,12 @@ def __call__(self, texts):
...
 class RearrangeSamples:
-    def __init__(self, n, wordlist):
-        self.n = n
+    def __init__(self, wordlist):
         self.wordlist = wordlist
 
     def __call__(self, items):
-        for i in range(len(items) // self.n):
-            yield [word for word in items[i*self.n: (i+1)*self.n] if word in self.wordlist]
+        yield [word for word in items if word in self.wordlist]
```
Comment (wowtor) [inline authorship/data.py]: If it is just to filter words, this class could be replaced by something simpler.

File: authorship/data.py [hunk 3]
```diff
@@ -40,15 +38,15 @@ def get(self):
         wordlist = [word for word, freq in get_frequent_words(speakers_wordlist, self._n_freqwords)]
 
         # build a dictionary of feature vectors
-        speakers = filter_texts_size(speakers_wordlist, wordlist, self._tokens_per_sample)
+        speakers_cov = filter_texts_size(speakers_wordlist, wordlist)
 
         # convert to X, y
-        X, y = to_vector_size(speakers)
+        X, y, ids = to_vector_size(speakers_cov)
```
Comment (wowtor) [inline authorship/data.py]: `ids` means speaker ids?

File: authorship/data.py [hunk 4]
```diff
@@ -40,15 +38,15 @@ def get(self):
         wordlist = [word for word, freq in get_frequent_words(speakers_wordlist, self._n_freqwords)]
 
         # build a dictionary of feature vectors
-        speakers = filter_texts_size(speakers_wordlist, wordlist, self._tokens_per_sample)
+        speakers_cov = filter_texts_size(speakers_wordlist, wordlist)
```
Comment (wowtor) [inline authorship/data.py]: cov is short for conversation?

File: authorship/data.py [hunk 5]
```diff
@@ -151,42 +152,45 @@ def get_frequent_words(speakers, n):
...
-    its values are the frequency of the words in the wordlist for that speaker/sample).
+    it returns one conversation of one speaker  (the length of the samples is equal to the length of the wordlist and
+    its values are the frequency of the words in the wordlist for that speaker/conversation).
 
     :param speakerdict: dict of all words used per speaker
     :param wordlist: the n most freq words in corpus
-    :param tokens_per_sample: number of words within a sample
     """
     filters = [
         ExtractWords(),
```
Comment (wowtor) [inline authorship/data.py]: `ExtractWords` doesn't seem to do anything. `RearrangeSamples` doesn't rearrange anything.

File: authorship/vocalize_data.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import numpy as np
+import pandas as pd
+
+
+class VocalizeDataSource:
```
Comment (wowtor) [inline authorship/vocalize_data.py:7]: Maybe a docstr here to describe the expected file format?

File: authorship/vocalize_data.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            endings = ('2-1', '4-1', '6-1', '8-1')
+        elif self._device == 'SM58close':
+            endings = ('2-2', '4-2')
+        elif self._device == 'AKGC400BL':
+            endings = ('2-3', '4-3')
+        elif self._device == 'SM58far':
+            endings = ('2-4', '4-4')
+        else:
+            endings = ''
+            print("vocalise: no device or incorrect device was given, no filter was applied")
```
Comment (wowtor) [inline authorship/vocalize_data.py]: does this actually happen? If it should NOT happen, replace the `print` by `raise`.

File: authorship/vocalize_data.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        df = df[df['SP_1'].str.endswith(endings) & df['SP_2'].str.endswith(endings)]
+
+        df[['SP_1', 'SP_2']] = df[['SP_1', 'SP_2']] \
+            .apply(lambda x: x.str.extract(conv_pattern[:(len(conv_pattern) - 7)] + ")", expand=False).str.replace("-", ""))
+
+        voc_pairs = df[['SP_1', 'SP_2']].to_numpy()
+        voc_pairs_set = np.apply_along_axis(set, 1, voc_pairs)
+        voc_score = df[['value']].to_numpy()
+
+        return (voc_pairs_set, np.array(voc_score))
```
Comment (wowtor) [inline authorship/vocalize_data.py]: It's hard for me to see if this works as intended, but if something goes wrong, results may be messed up and hard to debug. If you are sure that it works, fine. If unsure, unit tests would help.

File: authorship/vocalize_data.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import pandas as pd
+
+
+class VocalizeDataSource:
+    # specific on the vocalize output for frida
+    def __init__(self, voc_data, device='telephone'):
+        self._voc_data = voc_data
+        self._device = device
+
+    def get(self):
```
Comment (wowtor) [inline authorship/vocalize_data.py:23]: Maybe a docstr here to describe what it returns?

File: classify.py [hunk 10]
```diff
@@ -262,35 +263,47 @@ def __call__(self, lrs, y, title='', shortname=''):
         lir.ece.plot(lrs, y, path=ece_path, on_screen=not ece_path, kw_figure=kw_figure)
 
 
-def get_pairs(X, y, authors_subset, sample_size):
+def get_pairs(X, y, X_voc, conv_ids, authors_subset, sample_size):
```
Comment (wowtor) [inline classify.py]: docstr?

X = conversations
y = labels
X_voc = ?
conv_ids = conversation id per instance
authors_subset = labels to include
sample_size = maximum number of samples per label

actual number of samples is expected to be smaller if vocalize scores are missing, even if there is plenty data!

return: conversation pairs, labels 0 or 1, vocalize score

File: classify.py [hunk 11]
```diff
@@ -262,35 +263,47 @@ def __call__(self, lrs, y, title='', shortname=''):
...
-                                        different_source_limit=sample_size // 2).transform(X_subset, y_subset)
+    pairs_transformation = transformers.InstancePairing(same_source_limit=int(sample_size),
+                                                        different_source_limit=int(sample_size))
+    X_pairs, y_pairs = pairs_transformation.transform(X_subset, y_subset)
+    pairing_ids = pairs_transformation.pairing
+    con_pairs = np.apply_along_axis(lambda a: np.array([conv_ids_subset[a[0]], conv_ids_subset[a[1]]]), 1, pairing_ids)
 
+    # X_voc_set = np.apply_along_axis(set, 1, X_voc[0])
+    X_voc_set = X_voc[0]
+    voc_score = np.apply_along_axis(lambda a: X_voc[1][np.where(X_voc_set == set(a))[0]][0][0] if len(np.where(X_voc_set == set(a))[0]) == 1 else np.NaN, 1, con_pairs)
```
Comment (wowtor) [inline classify.py]: not sure what happens here

File: classify.py [hunk 12]
```diff
@@ -301,9 +314,18 @@ def evaluate_samesource(desc, dataset, n_frequent_words, tokens_per_sample, max_
     """
     # calibrator = lir.plotting.PlottingCalibrator(calibrator, lir.plotting.plot_score_distribution_and_calibrator_fit)
     clf = lir.CalibratedScorer(classifier, calibrator)
+    voc_cal = lir.ScalingCalibrator(lir.LogitCalibrator())
+    mfw_voc_clf = lir.CalibratedScorer(LogisticRegression(class_weight='balanced'), calibrator)
+    combine_features_flag = False
+    if len(clf.scorer.named_steps) > 1:
```
Comment (wowtor) [inline classify.py:349]: not sure what happens here

File: classify.py [hunk 13]
```diff
@@ -262,35 +263,47 @@ def __call__(self, lrs, y, title='', shortname=''):
...
 
 
-def evaluate_samesource(desc, dataset, n_frequent_words, tokens_per_sample, max_n_of_pairs_per_class, preprocessor, classifier, calibrator,
-                        plot=None, repeats=1):
+def evaluate_samesource(desc, dataset, voc_data, n_frequent_words, max_n_of_pairs_per_class, preprocessor, classifier,
+                        calibrator, plot=None, repeats=1):
     """
     Run an experiment with the parameters provided.
 
     :param desc: free text description of the experiment
```
Comment (wowtor) [inline classify.py:330]: docstr out of date

File: classify.py [hunk 14]
```diff
@@ -313,58 +335,121 @@ def evaluate_samesource(desc, dataset, n_frequent_words, tokens_per_sample, max_
     LOG.info(f'{desc}: number of speakers: {np.unique(y).size}')
     LOG.info(f'{desc}: number of instances: {y.size}')
 
-    X = preprocessor.fit_transform(X)
+    X = preprocessor.fit_transform(X) # shouldn't this take place on the x_train and then on x_test?
```
Comment (wowtor) [inline classify.py]: Is it possible to leak information from test data to training this way?

File: classify.py [hunk 15]
```diff
@@ -313,58 +335,121 @@ def evaluate_samesource(desc, dataset, n_frequent_words, tokens_per_sample, max_
...
 
     lrs = []
+    lrs_voc = []
+    lrs_comb = []
+    lrs_features = []
     y_all = []
-    for X_train, y_train, X_test, y_test in tqdm(get_batch_simple(X, y, repeats, max_n_of_pairs_per_class)):
-        # fit a classifier and calculate LRs
+
+    for X_train, y_train, X_test, y_test, X_voc_train, X_voc_test in tqdm(
```
Comment (wowtor) [inline classify.py]: confusing order: train, train, test, test, train, test
I would expect either train, train, train, test, test, test or train, test, train, test, train, test
What is the order used by sklearn train_test_split?

File: classify.py [hunk 16]
```diff
@@ -313,58 +335,121 @@ def evaluate_samesource(desc, dataset, n_frequent_words, tokens_per_sample, max_
...
+
+        # mfw - fit a classifier and calculate LRs (for m1, the resulted scorer can also be used for m2)
         clf.fit(X_train, y_train)
         lrs.append(clf.predict_lr(X_test))
         y_all.append(y_test)
 
+        # take scores from mfw scorer and combine with voc output using logit then calibrate (for m2)
+        mfw_scores_train = lir.apply_scorer(clf.scorer, X_train)
+        X_comb_train = np.vstack((np.squeeze(mfw_scores_train), X_voc_train)).T
+        mfw_voc_clf.fit(X_comb_train, y_train)
```
Comment (wowtor) [inline classify.py:408]: `mfw_scores_train` is a score in range 0..1
`X_voc_train` is an LR, log LR or something else?

Classifier may work better if both scores are in the same range, for example 0..1, or -1..1 (may require scaling)

File: classify.py [hunk 17]
```diff
@@ -313,58 +335,121 @@ def evaluate_samesource(desc, dataset, n_frequent_words, tokens_per_sample, max_
...
+        # NOTE: m3 can be used only if the scorer is a classification alg, otherwise m3 = m2 with scorer always a logit
+
+        # calculate LRs for vocalize output (for m1)
+        voc_cal.fit(X=X_voc_train, y=y_train)
+        lrs_voc.append(voc_cal.transform(X_voc_test))
+
+        # mfw - fit a classifier and calculate LRs (for m1, the resulted scorer can also be used for m2)
         clf.fit(X_train, y_train)
         lrs.append(clf.predict_lr(X_test))
         y_all.append(y_test)
```
Comment (wowtor) [inline classify.py:396]: suggested name change `lrs` to `lrs_mfw`

File: classify.py [hunk 18]
```diff
@@ -313,58 +335,121 @@ def evaluate_samesource(desc, dataset, n_frequent_words, tokens_per_sample, max_
...
+        X_comb_train = np.vstack((np.squeeze(mfw_scores_train), X_voc_train)).T
+        mfw_voc_clf.fit(X_comb_train, y_train)
+
+        mfw_scores_test = lir.apply_scorer(clf.scorer, X_test)
+        X_comb_test = np.vstack((np.squeeze(mfw_scores_test), X_voc_test)).T
+        lrs_comb.append(mfw_voc_clf.predict_lr(X_comb_test))
+
+        # check type of scorer (for m3). In current setting, a distance scorer always has 1 step while a ML has 2
+        if combine_features_flag:
+            X_train_onevector = clf.scorer.steps[0][1].transform(X_train)
```
Comment (wowtor) [inline classify.py:416]: not sure what happens here inside the steps[x][y]. May be sensitive to changes in the classifier setup

**Review by wowtor — COMMENTED** [review]
See inline comments

File: authorship.yaml (hunk 1 above)
Comment (elina-holmes) [reply authorship.yaml]: name change done. the file is 300mb, i prefer not to checked it in.

File: authorship/data.py (hunk 2 above)
Comment (elina-holmes) [reply authorship/data.py]: I removed both classes (RearrangeSamples and ExtractWords) as they made more sense in the previous setting

File: authorship/data.py (hunk 3 above)
Comment (elina-holmes) [reply authorship/data.py]: conversation id. I renamed it.

File: authorship/data.py (hunk 4 above)
Comment (elina-holmes) [reply authorship/data.py]: yes, i missed the n so i added it

File: authorship/data.py (hunk 5 above)
Comment (elina-holmes) [reply authorship/data.py]: both removed

File: authorship/vocalize_data.py (hunk 6 above)
Comment (elina-holmes) [reply authorship/vocalize_data.py:7]: done

File: authorship/vocalize_data.py (hunk 7 above)
Comment (elina-holmes) [reply authorship/vocalize_data.py]: replace print with raise -> done

File: authorship/vocalize_data.py (hunk 9 above)
Comment (elina-holmes) [reply authorship/vocalize_data.py:23]: done

File: classify.py (hunk 10 above)
Comment (elina-holmes) [reply classify.py]: done

File: classify.py (hunk 11 above)
Comment (elina-holmes) [reply classify.py]: comments were added

File: classify.py (hunk 12 above)
Comment (elina-holmes) [reply classify.py:349]: comment was added

File: classify.py (hunk 13 above)
Comment (elina-holmes) [reply classify.py:330]: done

File: classify.py (hunk 15 above)
Comment (elina-holmes) [reply classify.py]: changed to [train sets], [test sets]

File: classify.py (hunk 17 above)
Comment (elina-holmes) [reply classify.py:396]: done

File: classify.py (hunk 16 above)
Comment (elina-holmes) [reply classify.py:408]: done

File: classify.py (hunk 14 above)
Comment (elina-holmes) [reply classify.py]: done
